drivers: sensor: add INA700 power and current sensor driver

aunsbjerg opened the pull request:
Texas Instruments INA700 power and current sensor driver.

Based on existing drivers for INA230 and others. 

Currently supports both trigger and non-trigger configurations, but trigger configuration only enabled for conversion ready interrupt and only using system workqueue.

Tested on out-of-tree custom board.

Commit:
drivers: sensor: add ina700 power and current sensor

The ina700 sensor driver is inspired by the existing ina23x sensor
drivers as there is a fair bit of overlap between the devices.

The ina700 sensor driver supports both trigger and non-trigger variants,
with the trigger variant being limited to only work on the system
workqueue and only for conversion ready interrupts from the alert gpio.

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@escolifesciences.com>

Commit:
tests: drivers: build_all: sensor: add ina700 sensor

For both trigger and non-trigger variants.

Signed-off-by: Mikkel Jakobsen <mikkel.aunsbjerg@escolifesciences.com>

jilaypandya commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -53,6 +53,7 @@ add_subdirectory_ifdef(CONFIG_MHZ19B mhz19b)
 add_subdirectory_ifdef(CONFIG_NCT75 nct75)
 add_subdirectory_ifdef(CONFIG_NTC_THERMISTOR ntc_thermistor)
 add_subdirectory_ifdef(CONFIG_PMS7003 pms7003)
+add_subdirectory_ifdef(CONFIG_INA700 ina700)
```
Should be after HP206C

aunsbjerg replied:
I moved this to the `ti` subfolder which I missed in the first round.

jilaypandya commented on drivers/sensor/ina700/ina700.c:
```diff
@@ -0,0 +1,333 @@
+	const struct ina700_config *cfg = dev->config;
+	uint8_t rx_buf[3];
+
+	int rc = i2c_burst_read_dt(&cfg->bus, addr, rx_buf, sizeof(rx_buf));
```
```suggestion
	int rc = i2c_burst_read_dt(&cfg->bus, addr, rx_buf, sizeof(rx_buf));
	if (rc) {
	    return rc;
	}
```

jilaypandya commented on drivers/sensor/ina700/ina700.c:
```diff
@@ -0,0 +1,333 @@
+
+	if ((chan == SENSOR_CHAN_VOLTAGE) || (chan == SENSOR_CHAN_ALL)) {
+		rc = reg_read_16(dev, INA700_REG_VBUS, &data->voltage);
+
```
```suggestion

```

aunsbjerg replied:
I much prefer an empty line before if statements - to me it's more readable. Is this a code standard thing? Otherwise, I will leave it as is ;)

jilaypandya replied:
Across the codebase you would see it like this:
```

err = do_something();
if (err) {
        return err;
}
```

Hence, i suggested to drop that newline :)

jilaypandya commented on drivers/sensor/ina700/ina700.c:
```diff
@@ -0,0 +1,333 @@
+static int ina700_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	int rc = 0;
```
Not assigned before being read :)
```suggestion
	int rc;
```

jilaypandya commented on drivers/sensor/ina700/ina700.c:
```diff
@@ -0,0 +1,333 @@
+		}
+	}
+
+	ina700_unlock_data(data->this);
```
flatten out nested-ifs :)
```suggestion
	if (reg_read_16(data->this, INA700_REG_ALERT_DIAGNOSTICS, &status) != 0) {
	        goto unlock;
	}    
	if (!(status & INA700_ALERT_DIAG_CONVERSION_READY)) {
	        goto unlock;
	}    
	if (ina700_sample_fetch(data->this, SENSOR_CHAN_ALL) != 0) {
	        goto unlock;
	}    
	if (data->handler_alert) {
		data->handler_alert(data->this, data->trigger_alert);
	}
unlock:
	ina700_unlock_data(data->this);
```

jilaypandya commented on drivers/sensor/ina700/ina700.c:
```diff
@@ -0,0 +1,333 @@
+	ARG_UNUSED(dev);
+}
+
+static void ina700_unlock_data(const struct device *dev)
```
I think you can drop these wrappers and use mutex_lock/mutex_unlock directly within functions :)

aunsbjerg replied:
Then we would need to wrap each call to mutex_lock/unlock in an ifdef block - the mutex member is only defined if the trigger is enabled through kconfig.

jilaypandya replied:
Yeah but why not use mutex each time you access the data, probably its not done in all the drivers, but i think it would be completely alright. 

Or atleast have just one function definition and you can do ifdefery within that function :)

jilaypandya commented on drivers/sensor/ina700/ina700.c:
```diff
@@ -0,0 +1,333 @@
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
```
Some of these includes are already present in ina700.h :)

aunsbjerg commented:
@jilaypandya thanks for the review, I appreciate it. I believe I have addressed your comments. Also had to rebase because of a conflict with a PR that was merged in the meantime.
